lib/message/context.ts

```diff
@@ -0,0 +1,102 @@
+/**
+ * The context is an abstraction we use to add domain separation to OpenPGP signatures.
+ * It's implemented by adding notation data to signatures, which may be marked as critical, so that
+ * the resulting signature can only be verified by a verifier who expects the context to be present.
+ */
+import { SignaturePacket, RawNotation, PartialConfig } from '../openpgp';
+import { normalizeDate } from './utils';
+
+// Different contexts will affect the notation's value, not the name
+const CONTEXT_NOTATION_NAME = 'context@proton.ch';
+
+export interface ContextSigningOptions {
+    /** context identifier */
+    id: string,
+    /**
+     * Whether verification must fail unless the verifier is context-aware:
+     * if `true`, the signature cannot be verified unless the verifier expects the same context.
+     * If `false`, the signature will verify successfully even if the verifier does not expect any context.
+     * This should be set to `true` as long as all clients verifying the signature are able to process contexts.
+     */
+    critical: boolean
+}
+
+export type ContextVerificationOptions = {
+    id: string,
+    /**
+     * Whether the signature must include context info for verification to succeed:
+     * if `true`, a signature with no context won't verify. If `false`, it will.
+     * Note: if the context is not required, but a different context is found in the signature, verification will always fail.
+     */
+    required: boolean,
+    requiredFrom?: undefined
+} | {
+    id: string,
+    required?: undefined
+    /**
+     * Signatures created after this date must include context info for verification to succeed (same as `expect: true`).
+     * For signatures created before the given date, the verification behavior is equivalent to `expect: false`
+     * */
+    requiredFrom: Date,
+};
+
+/**
+ * Translate a `contextID` string into an OpenPGP notation object, which can be signed as part of the message
+ * @param contextID - context identifier
+ * @param critical - whether the notation should be critical (if so, verification will fail if in the wrong context,
+ *  i.e. if the caller did not declare the notation as known)
+ */
+export const getNotationForContext = (contextID: string, critical: boolean): RawNotation => ({
+    name: CONTEXT_NOTATION_NAME,
+    value: new TextEncoder().encode(contextID),
+    humanReadable: true,
+    critical
+});
+
+/**
+ * Confirm context validity by finding the corresponding signature notation.
+ * NB: signature validity is not checked.
+ */
+export const isValidSignatureContext = (contextOptions: ContextVerificationOptions, signature: SignaturePacket) => {
+    const { id: contextID, required, requiredFrom } = contextOptions;
+    const isContextRequired = requiredFrom ? normalizeDate(requiredFrom) <= signature.created! : !!required;
```

Comment:
And then this reads slightly more naturally as
```suggestion
    const { id: contextID, required, requiredAfter } = contextOptions;
    const isContextRequired = requiredAfter ? signature.created! > normalizeDate(requiredAfter) : !!required;
```